pkg/mcs/discovery/discover.go

```diff
@@ -75,5 +78,68 @@ func GetMSMembers(name string, client *clientv3.Client) ([]ServiceRegistryEntry,
 		return entries, nil
 	}
 
-	return nil, errors.Errorf("unknown service name %s", name)
+	return nil, errors.Errorf("unknown service name %s", serviceName)
+}
+
+// TransferPrimary transfers the primary of the specified service.
+func TransferPrimary(client *clientv3.Client, serviceName, oldPrimary, newPrimary string, keyspaceGroupID uint32) error {
+	log.Info("transfer primary", zap.String("service", serviceName), zap.String("from", oldPrimary), zap.String("to", newPrimary))
+	entries, err := GetMSMembers(serviceName, client)
+	if err != nil {
+		return err
+	}
+
+	// Do nothing when I am the only member of cluster.
+	if len(entries) == 1 {
+		return errors.New("no valid secondary to transfer primary")
+	}
+
+	var primaryIDs []string
+	var secondaryValues []string
+	for _, member := range entries {
+		if (newPrimary == "" && member.ServiceAddr != oldPrimary) || (newPrimary != "" && member.Name == newPrimary) {
+			primaryIDs = append(primaryIDs, member.ServiceAddr)
+			if string(member.MemberValue) == "" {
+				return errors.New("member value is empty")
+			}
+			secondaryValues = append(secondaryValues, string(member.MemberValue))
+		}
+	}
+	if len(primaryIDs) == 0 {
+		return errors.New("no valid secondary to transfer primary")
+	}
+
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	nextPrimaryID := r.Intn(len(primaryIDs))
+
+	clusterID, err := etcdutil.GetClusterID(client, utils.ClusterIDPath)
+	if err != nil {
+		return errors.Errorf("failed to get cluster ID: %v", err)
+	}
+
+	var primaryKey string
+	switch serviceName {
+	case utils.SchedulingServiceName:
+		primaryKey = endpoint.SchedulingPrimaryPath(clusterID)
+	case utils.TSOServiceName:
+		tsoRootPath := endpoint.TSOSvcRootPath(clusterID)
+		primaryKey = endpoint.KeyspaceGroupPrimaryPath(tsoRootPath, keyspaceGroupID)
+	}
+
+	// remove possible residual value
+	utils.RemoveExpectedPrimary(client, primaryKey)
+
+	// only give the primary lease to the new primary
+	grantResp, err := client.Grant(client.Ctx(), utils.DefaultLeaderLease)
+	if err != nil {
+		return errors.Errorf("failed to grant lease for %s, err: %v", serviceName, err)
+	}
+	// update primary key to notify old primary server.
+	putResp, err := kv.NewSlowLogTxn(client).
```

Comment:
If the transaction takes a long time here, will it be possible to make the lease expire unexpectedly?

Reply:
We give lease 3 seconds.
- If this transaction takes a long time will only have the effect of causing the transfer primary to start very slowly, which will also be reflected by the api take a long time to execute it.
- If you mean that the grant lease was not updated due to a large transaction failure, then the primary will be re-selected after a timeout. You can checked in `TestTransferPrimaryWhileLeaseExpired`

https://github.com/tikv/pd/pull/8157/files#diff-5a1f6bff4bb6b99fcd2f7bf4edc37aa937e8c51eafa41acc0aabd7aa1a1d1da6R267